js/compiler.ts

```diff
@@ -0,0 +1,586 @@
+      throw new TypeError("Attempt to create an additional compiler.");
+    }
+    this._service = this._ts.createLanguageService(this);
+    this._setupSourceMaps();
+  }
+
+  // Deno specific compiler API
+
+  /**
+   * Retrieve the output of the TypeScript compiler for a given `fileName`.
+   */
+  compile(fileName: ModuleFileName): OutputCode {
+    const service = this._service;
+    const output = service.getEmitOutput(fileName);
+
+    // Get the relevant diagnostics - this is 3x faster than
+    // `getPreEmitDiagnostics`.
+    const diagnostics = [
+      ...service.getCompilerOptionsDiagnostics(),
+      ...service.getSyntacticDiagnostics(fileName),
+      ...service.getSemanticDiagnostics(fileName)
+    ];
+    if (diagnostics.length > 0) {
+      const errMsg = this._ts.formatDiagnosticsWithColorAndContext(
+        diagnostics,
+        formatDiagnosticsHost
+      );
+      console.log(errMsg);
+      // All TypeScript errors are terminal for deno
+      this._os.exit(1);
+    }
+
+    assert(!output.emitSkipped, "The emit was skipped for an unknown reason.");
+
+    // Currently we are inlining source maps, there should be only 1 output file
+    // See: https://github.com/denoland/deno/issues/23
+    assert(
+      output.outputFiles.length === 1,
+      "Only single file should be output."
+    );
+
+    const [outputFile] = output.outputFiles;
+    return outputFile.text;
+  }
+
+  /**
+   * Create a localized AMD `define` function and return it.
+   */
+  makeDefine(moduleMetaData: ModuleMetaData): AmdDefine {
+    const localDefine = (deps: string[], factory: AmdFactory): void => {
+      // TypeScript will emit a local require dependency when doing dynamic
+      // `import()`
+      const localRequire: AmdRequire = (
+        deps: string[],
+        callback: AmdCallback,
+        errback?: AmdErrback
+      ): void => {
+        this._log("localRequire", deps);
+        try {
+          const args = deps.map(dep => {
+            if (dep in DenoCompiler._builtins) {
+              return DenoCompiler._builtins[dep];
+            } else {
+              const depModuleMetaData = this.run(dep, moduleMetaData.fileName);
+              return depModuleMetaData.exports;
+            }
+          });
+          callback(...args);
+        } catch (e) {
+          if (errback) {
+            errback(e);
+          } else {
+            throw e;
+          }
+        }
+      };
+      const localExports = moduleMetaData.exports;
+      this._log("localDefine", moduleMetaData.fileName, deps, localExports);
+      const args = deps.map(dep => {
+        if (dep === "require") {
+          return localRequire;
+        } else if (dep === "exports") {
+          return localExports;
+        } else if (dep in DenoCompiler._builtins) {
+          return DenoCompiler._builtins[dep];
+        } else {
+          const depModuleMetaData = this.run(dep, moduleMetaData.fileName);
+          return depModuleMetaData.exports;
```

Comment:
This code is duplicated just above in the try-catch? I'm having trouble following the logic here. Is this a programming error? If not, why are we mapping the deps twice?

Reply:
Not done. @kitsonk please take a look at this when you get a chance.